File: problems/lowest_common_ancestor_of_a_binary_tree/lowest_common_ancestor_of_a_binary_tree.cc
// Unsolved. Need more practice.

#include <iostream>

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
 public:
  TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {    
    // Return the lowest common ancestor of a binary tree.
    if (!root || root == p || root == q) {
      return root;
    }
    TreeNode* left = lowestCommonAncestor(root->left, p, q);
    TreeNode* right = lowestCommonAncestor(root->right, p, q);
    // If both left and right have values, means the result is root.
    // return !left ? right : !right ? left : root;
    if (!left) {
      return right;
    }
    if (!right) {
      return left;
    }
    return root;
  }
  
  // A Contain function is unnecessary
  // bool Contain(TreeNode* root, TreeNode* p) {
  //   if (root->val == p->val) {
  //     return true;
  //   }
  //   if (root->left && Contain(root->left, p)) {
  //     return true;
  //   }
  //   if (root->right && Contain(root->right, p)) {
  //     return true;
  //   }
  //   return false;
  // }
};

int main(int argc, char** argv) {
  //        -1
  //      0     3
  //   -2    4
  // 8
  TreeNode nodes[6] = {-1, 0, 3, -2, 4, 8};
  TreeNode* root = &nodes[0];
  root->left = &nodes[1];
  root->right = &nodes[2];
  root->left->left = &nodes[3];
  root->left->right = &nodes[4];
  root->left->left->left = &nodes[5];
  
  Solution solution = Solution();
  
  TreeNode* node = solution.lowestCommonAncestor(root, root->left->left->left, root->left->right);
  std::cout << "Print the lowestCommonAncestor." << std::endl;
  std::cout << node->val << std::endl;
  
  return 0;
}

File: problems/lowest_common_ancestor_of_a_binary_tree/lowest_common_ancestor_of_a_binary_tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        # If root is either p or q, it is the part of answer.
        if root in (None, p, q):
            return root

        left, right = (self.lowestCommonAncestor(child, p, q)
                       for child in (root.left, root.right))

        # The LCA is the root has p and q in different side.
        # If p and q are in the same side, the LCA is in that side.
        return root if left and right else left or right
